source/common/network/dns_resolver/dns_factory.h

```diff
@@ -0,0 +1,146 @@
+            config.typed_dns_resolver_config(), true);
+    if ((dns_resolver_factory != nullptr) &&
+        (dns_resolver_factory->category() == std::string(DnsResolverCategory))) {
+      typed_dns_resolver_config.MergeFrom(config.typed_dns_resolver_config());
+      return true;
+    }
+  }
+  return false;
+}
+
+// If the config has dns_resolution_config, create a CaresDnsResolverConfig typed config based on
+// it.
+template <class ConfigType>
+bool checkDnsResolutionConfigExist(
+    const ConfigType& config,
+    envoy::config::core::v3::TypedExtensionConfig& typed_dns_resolver_config) {
+  if (config.has_dns_resolution_config()) {
+    envoy::extensions::network::dns_resolver::cares::v3::CaresDnsResolverConfig cares;
+    if (!config.dns_resolution_config().resolvers().empty()) {
+      cares.mutable_resolvers()->MergeFrom(config.dns_resolution_config().resolvers());
+    }
+    cares.mutable_dns_resolver_options()->MergeFrom(
+        config.dns_resolution_config().dns_resolver_options());
+    typed_dns_resolver_config.mutable_typed_config()->PackFrom(cares);
+    typed_dns_resolver_config.set_name(std::string(CaresDnsResolver));
+    return true;
+  }
+  return false;
+}
+
+// For backward compatibility, copy over use_tcp_for_dns_lookups from config, and create
+// a CaresDnsResolverConfig typed config. This logic fit for bootstrap, and dns_cache config types.
+template <class ConfigType>
+void handleLegacyDnsResolverData(
+    const ConfigType& config,
+    envoy::config::core::v3::TypedExtensionConfig& typed_dns_resolver_config) {
+  envoy::extensions::network::dns_resolver::cares::v3::CaresDnsResolverConfig cares;
+  cares.mutable_dns_resolver_options()->set_use_tcp_for_dns_lookups(
+      config.use_tcp_for_dns_lookups());
+  typed_dns_resolver_config.mutable_typed_config()->PackFrom(cares);
+  typed_dns_resolver_config.set_name(std::string(CaresDnsResolver));
+}
+
+// Overloading the template function for DnsFilterConfig type, which doesn't need to copy anything.
+void handleLegacyDnsResolverData(
+    const envoy::extensions::filters::udp::dns_filter::v3alpha::DnsFilterConfig::
+        ClientContextConfig&,
+    envoy::config::core::v3::TypedExtensionConfig& typed_dns_resolver_config);
+
+// Overloading the template function for Cluster config type, which need to copy
+// both use_tcp_for_dns_lookups and dns_resolvers.
+void handleLegacyDnsResolverData(
+    const envoy::config::cluster::v3::Cluster& config,
+    envoy::config::core::v3::TypedExtensionConfig& typed_dns_resolver_config);
+
+// Retrieve the DNS related configurations in the passed in @param config, and store the data into
+// @param typed_dns_resolver_config.
+template <class ConfigType>
+void makeDnsResolverConfig(
+    const ConfigType& config,
+    envoy::config::core::v3::TypedExtensionConfig& typed_dns_resolver_config) {
+  // typed_dns_resolver_config takes precedence
+  if (checkTypedDnsResolverConfigExist(config, typed_dns_resolver_config)) {
```

Comment:
If someone specifies a typed_dns_resolver_config directly, and it's invalid, I think we should fail. I don't think we should fall back. Am I misunderstanding this code? I would just let it fail/throw out.

Reply:
HI, @mattklein123 

Thanks for the comments.

Yes, we had thought about what's the right behavior when an invalid typed_dns_resolver_config is in the config. The reason we chose to fall back to create default resolver ( c-ares in all cases or Apple if in MacOS + run_time_use_apple_dns enabled ) is because:

1) If somehow control plane sends an incorrect typed_dns_resolver_config by xDS, always throw exception/fail Envoy instance here make it vulnerable. 
2) The current logic always creates a resolver in this context: https://github.com/envoyproxy/envoy/blob/b3f06c4b67934a9f3e637cf99c34b980fb82ac04/source/common/event/dispatcher_impl.cc#L160. There is no logic to deal with wrong configuration in this context or return a nullptr.  Thus I am trying to maintain that logic.

But, I do agree silently fall back to default maybe not right.  Is adding a warning message like below acceptable?

diff --git a/source/common/network/dns_resolver/dns_factory.h b/source/common/network/dns_resolver/dns_factory.h
index 3dfd73d3c..f89b03498 100644
--- a/source/common/network/dns_resolver/dns_factory.h
+++ b/source/common/network/dns_resolver/dns_factory.h
@@ -67,6 +67,11 @@ bool checkTypedDnsResolverConfigExist(
       typed_dns_resolver_config.MergeFrom(config.typed_dns_resolver_config());
       return true;
     }
+    ENVOY_LOG_MISC(warn, "Warning: Either the typed_dns_resolver_config {} is not registered"
+                   " or the configuration with type URL {} is not a DNS resolver type."
+                   " Fall back to creating default DNS resolver.",
+                   config.typed_dns_resolver_config().name(),
+                   config.typed_dns_resolver_config().typed_config().type_url());
   }
   return false;
 }

If this is not sufficient, please advise what else we should add.

Thanks!

Reply:
What I would recommend doing is pre-verifying the typed_config for DNS when the config is actually loading (so do it manually for cluster and in the other places). Then you can fail/throw the config with a good message. Then on the actual creation part you can know that it is correct.

Reply:
HI, @mattklein123 ,
After further checking, we found we can actually utilizing below registry lookup code in source/common/event/dispatcher_impl.cc to fail/throw an exception if the typed DNS resolver config is bad:

  // Derive the DNS resolver factory from config.
  // If the typed config is not registered or bogus, it will throw an exception.
  // Since this function is only called in control plane, it is safe to throw an exception here.
  Network::DnsResolverFactory& dns_resolver_factory =
      Config::Utility::getAndCheckFactory<Network::DnsResolverFactory>(typed_dns_resolver_config);

The same check(with is_optional false) in checkTypedDnsResolverConfigExist() is not needed, and is removed.

Since dispatcherimpl::createDnsResolver() is always called in control plane code (either during Envoy initialization or failter creation), throw an exception in these cases are fine.

I added two unit test cases in dispatcher_impl_test.cc to cover good/bad typed_dns_resolver_config passed in this function. PTAL.

thanks!